wormhole_attester/program/src/attest.rs

```diff
@@ -224,53 +218,70 @@ pub fn attest(ctx: &ExecutionContext, accs: &mut Attest, data: AttestData) -> So
+            // Used as last_attested_publish_time, defaults to the new_* value if no pre-existing state is available
+            state
+                .0
+                 .1
+                .last_attested_trading_publish_time
+                .unwrap_or(new_last_attested_trading_publish_time),
             price_struct,
         );
 
-
-        // update last_attested_publish_time with this price's
-        // publish_time. Yes, it may be redundant for the entry() used
-        // above in the rare first attestation edge case.
-        state_entry.last_attested_trading_publish_time = last_trading_publish_time;
-
-        // The following check is crucial against poorly ordered
-        // account inputs, e.g. [Some(prod1), Some(price1),
-        // Some(prod2), None, None, Some(price)], interpreted by
-        // earlier logic as [(prod1, price1), (prod2, price3)].
-        //
-        // Failing to verify the product/price relationship could lead
-        // to mismatched product/price metadata, which would result in
-        // a false attestation.
-        if attestation.product_id.to_bytes() != product.key.to_bytes() {
-            trace!(&format!(
-                "Price's product_id does not match the pased account (points at {:?} instead)",
-                attestation.product_id
-            ));
-            return Err(ProgramError::InvalidAccountData.into());
+        // Update the on-chain value using publish_time or
+        // prev_publish_time if the price is not currently trading
+        state.0.last_attested_trading_publish_time = Some(new_last_attested_trading_publish_time);
+
+        // Serialize the state to calculate rent/account size adjustments
+        let state_serialized = state.0 .1.try_to_vec()?;
+
+        if state.0.is_initialized() {
+            state.0.info().realloc(state_serialized.len(), false)?;
+            trace!("Attestation state resize OK");
+
+            let target_rent = CreationLamports::Exempt.amount(state_serialized.len());
+            let current_rent = state.0.info().lamports();
+
+            // Adjust rent, but only if there isn't enough
+            if target_rent > current_rent {
+                let transfer_amount = target_rent - current_rent;
+
+                let transfer_ix = system_instruction::transfer(
+                    accs.payer.info().key,
+                    state.0.info().key,
+                    transfer_amount,
+                );
+
+                invoke(&transfer_ix, ctx.accounts)?;
+            }
+
+            trace!("Attestation state rent transfer OK");
+        } else {
+            let seeds = state.self_bumped_seeds(price.key, ctx.program_id);
+            solitaire::create_account(
+                ctx,
+                state.0.info(),
+                accs.payer.key,
+                solitaire::CreationLamports::Exempt,
+                state_serialized.len(),
```

Comment:
Instead of use `state_serialized`, create a constant `AttestationState::LEN` which is equal to the maximum size the account can have. This will save you resizing.

Reply:
Or turn `last_attested_trading_publish_time` into a `Timestamp` instead of an Option and use the `AccountSize` solitaire trait.

Reply:
I think I prefer spending the extra gas in exchange for not maintaining a constant for this. I think without the realloc logic it can live in the uninitialized if branch, so this is only done once in a PDA's lifetime ( that's the case in the change that removes the unneeded realloc()). Fundamentally, I think Solitaire could benefit from a flag that triggers this call for uninitialized data accounts. On persist() it serializes the struct anyway.